### PR TITLE
Convert -private/utils to TypeScript

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
this directory is just one file. figured I'd start small :)

### CODE REVIEW

File: addon/-private/utils/parse-response-headers.ts [hunk 1]
```diff
@@ -1,7 +1,11 @@
 const newline = /\r?\n/;
 
-export default function parseResponseHeaders(headersString) {
-  let headers = Object.create(null);
+export type ResponseHeaders =
+| {}
+| { [key: string]: string };
```
Comment (mike-north) [inline addon/-private/utils/parse-response-headers.ts]: Why have you chosen this type instead of ` { [key: string]: string } `?

Comment (NullVoxPopuli) [reply addon/-private/utils/parse-response-headers.ts]: good question, I don't remember. everything seems fine without the extra `{}`

File: addon/-private/utils/parse-response-headers.ts [hunk 2]
```diff
@@ -1,7 +1,11 @@
 const newline = /\r?\n/;
 
-export default function parseResponseHeaders(headersString) {
-  let headers = Object.create(null);
+interface ResponseHeaders {
+  [key: string]: string
+};
```
Comment (mike-north) [inline addon/-private/utils/parse-response-headers.ts]: This type is way more foundational than just "response headers". 

Let's create a `..../-private/types.ts` and have a

```ts
export interface Dict<T = unknown> {
   [k: string]: T | undefined;
}
```

and then you can import it and

```ts
export default function parseResponseHeaders(headersString: Dict<string>)
```

Comment (NullVoxPopuli) [reply addon/-private/utils/parse-response-headers.ts]: would that not be something you'd want in the global index.d.ts?

Comment (mike-north) [reply addon/-private/utils/parse-response-headers.ts]: Nope. Consumers of the ember-data won't get any of those types.

Comment (mike-north) [reply addon/-private/utils/parse-response-headers.ts]: I'm not sure you understand. Everything in your `/types` folder is for _your project only_. If you put `Dict` there, and used it anywhere in parts of the library that users of ember-data touch, it'll cause them problems.

Comment (NullVoxPopuli) [reply addon/-private/utils/parse-response-headers.ts]: right, in my latest commit, I moved the Dict type to -private/types.ts, and it didn't need to be imported. :-\

File: addon/-private/types.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+type Dict<T = unknown> = {
```
Comment (mike-north) [inline addon/-private/types.ts]: you still need the export, and the corresponding import

**Comment by runspired** [comment]
Thanks for the effort! I think this PR is stale enough likely we should reboot from where we are now.

We have a `Dict` in our `ts-interfaces` currently but the upgrades listed here would be welcome.

**Pull request closed without merging**